### ISSUE TITLE
type-narrowing involving `all` and `filter`

### ISSUE THREAD

**Issue by NCPlayz** [microsoft/pyright#2012]
**Is your feature request related to a problem? Please describe.**
Currently, pyright does not seem to utilise `all` and `filter` to narrow down the type of an iterable.
```py
from typing import List, Optional

def test():
    a: List[Optional[str]] = []

    if all(item is not None for item in a):
        return

    reveal_type(a)  # Type of "a" is "list[str | None]"

    b = list(filter(lambda i: i is not None, a))

    reveal_type(b)  # Type of "b" is "list[str | None]"
```
The type of `a` and `b` should ideally be `list[str]` as the possibility of any item being `None` has been nullified, so these two builtins may need to be special-cased.

**Describe the solution you'd like**
For `filter`, it seems that the (list) comprehension equivalent works fine:
```py
    b = list(filter(lambda item: item is not None, a))  # Type of "b" is "list[str | None]"

    reveal_type(b)

    c = [item for item in a if item is not None]  # Type of "b" is "list[str]"

    reveal_type(c)
```
So, I wonder if `filter(x, y)` could use the same logic as `i for i in y if x(i)`?

**Additional context**
While mypy does result in the same revealed types as pyright, TypeScript's `Array.every` and `Array.filter` offer the same interfaces and correctly narrow down the type, so it seems to make logical sense to support this behaviour in this case.

**Comment by JelleZijlstra** [comment]
For `filter`, this should be solved once we amend typeshed to use PEP 647 (TypeGuard) in its definition. I just opened python/typeshed#5661 for that.

`all()` is harder because it cannot be done just in stubs; it will require changes to mypy itself.

**Comment by erictraut** [comment]
We generally don't hard-code knowledge about the semantics of specific functions within pyright. That approach simply isn't scalable. Type-related knowledge should come from type stubs.

As @JelleZijlstra said, `filter` can be annotated using `TypeGuard`.

I don't see a good way to generalize `all`. The type checker would not only need to hard code knowledge of `all` but also make assumptions about the semantics of the iterable expression it is acting upon. For example, let's say that the iterable expression was of type `Foo[int | None]`, and `Foo` happened to implement an `__iter__` method, making it conform to the `Iterable` protocol. Does that mean that it can be narrowed to `Foo[int]` if you use it within an `all` expression? No, it wouldn't be safe to make that assumption.

TypeScript is able to handle this because `Array.every` acts only on `Array` which has well-understood semantics. `all` acts upon any `Iterable`.

I think the best option for `all` is for you to write your own user-defined type guard function which is an equivalent of all. Here's what that would look like:

```python
from typing import List, Optional, TypeGuard, TypeVar

T = TypeVar("T")

def is_list_without_nones(a: List[Optional[T]]) -> TypeGuard[List[T]]:
    return all(item is not None for item in a)

def test(a: List[Optional[str]]):
    if is_list_without_nones(a):
        reveal_type(a)  # Type of "a" is "list[str]"
```
